pkg/proxy/healthcheck/healthcheck.go

```diff
@@ -0,0 +1,137 @@
+	serviceName types.NamespacedName
+	endpoints   *sets.String
+}
+
+// serviceResponder: Contains net/http datastructures necessary for responding to each Service's health check on its aux nodePort
+type serviceResponder struct {
+	serviceName types.NamespacedName
+	listenPort  uint16
+	listener    *net.Listener
+	server      *http.Server
+}
+
+// proxyHC: Handler structure for health check, endpoint add/delete and service listener add/delete requests
+type proxyHC struct {
+	serviceEndpointsMap    cache.ThreadSafeStore
+	serviceResponderMap    map[types.NamespacedName]serviceResponder
+	mutationRequestChannel chan *proxyMutationRequest
+	listenerRequestChannel chan *proxyListenerRequest
+	shutdownChannel        chan bool
+}
+
+// handleHealthCheckRequest - received a health check request - lookup and respond to HC.
+func (h *proxyHC) handleHealthCheckRequest(rw http.ResponseWriter, serviceName string) {
+	s, ok := h.serviceEndpointsMap.Get(serviceName)
+	if !ok {
+		glog.V(3).Infof("Service %s not found or has no local endpoints", serviceName)
+		sendHealthCheckResponse(rw, http.StatusServiceUnavailable, "No Service Endpoints Not Found")
+		return
+	}
+	service := s.(serviceEndpointsList)
+	numEndpoints := len(*service.endpoints)
+	glog.V(3).Infof("Service %s %d endpoints found", serviceName, numEndpoints)
+	if numEndpoints > 0 {
+		sendHealthCheckResponse(rw, http.StatusOK, fmt.Sprintf("%d Service Endpoints found", numEndpoints))
+		return
+	}
+	sendHealthCheckResponse(rw, http.StatusServiceUnavailable, "0 local Endpoints are alive")
+}
+
+// handleMutationRequest - receive requests to mutate the table entry for a service
+func (h *proxyHC) handleMutationRequest(req *proxyMutationRequest) {
+	numEndpoints := len(*req.endpointUids)
+	glog.V(3).Infof("LB service health check mutation request Service: %s - %d Endpoints %v",
+		req.serviceName, numEndpoints, req.endpointUids.List())
+	switch {
+	case numEndpoints == 0:
+		if _, ok := h.serviceEndpointsMap.Get(req.serviceName.String()); ok {
+			glog.V(4).Infof("Deleting endpoints map for service %s, all local endpoints gone", req.serviceName.String())
+			h.serviceEndpointsMap.Delete(req.serviceName.String())
+		}
+
+	case numEndpoints > 0:
+		var entry serviceEndpointsList
+		e, ok := h.serviceEndpointsMap.Get(req.serviceName.String())
+		if !ok {
+			glog.V(2).Infof("Creating new endpoints map for service %s", req.serviceName.String())
+			emptyset := sets.NewString()
+			entry = serviceEndpointsList{serviceName: req.serviceName, endpoints: &emptyset}
+		} else {
+			entry = e.(serviceEndpointsList)
+		}
+		if !entry.endpoints.Equal(*req.endpointUids) {
+			// Compute differences just for printing logs about additions and removals
+			deletedEndpoints := entry.endpoints.Difference(*req.endpointUids)
+			newEndpoints := req.endpointUids.Difference(*entry.endpoints)
+			for _, e := range newEndpoints.List() {
+				glog.V(2).Infof("Adding local endpoint %s to LB health check for service %s",
```

Comment:
want to log this at V(4)? its going to be noisy for every add/delete endpoint. V(4) will show up in tests but V(2) will show up in customer logs. Choose one, depending on what you plan to use this for.